File: StriverA2Z/C++/3_ArraysQuestion/3.2_Medium/majority_element_n_by_two.cpp
#include<iostream>

using namespace std;

void majorityElementNByTwo(int arr[], int n) {
    int el;
    int count = 0;
    for (int i = 0; i < n; i++) {
        if (count == 0) {
            count = 1;
            el = arr[i];
        } else if (arr[i] == el) {
            count++;
        } else if (arr[i] != el) {
            count--;
        }
    }

    int cnt = 0;
    for (int i = 0; i < n; i++) {
        if (arr[i] == el) cnt++;
    }

    if (cnt > n / 2) {
        cout << el << endl;
    } else {
        cout << -1;
    }
}

int main() {
    int arr[] = {2, 2, 1, 3, 1, 1, 3, 1, 1};
    majorityElementNByTwo(arr, sizeof(arr) / sizeof(arr[0]));
    return 0;
}

File: StriverA2Z/Java/Arrays/Medium/MajorityElementNByTwo.java
public class MajorityElementNByTwo {
    static void majorityElement(int[] arr) {
        int el = arr[0];
        int count = 0;
        for (int k : arr) {
            if (count == 0) {
                count = 1;
                el = k;
            } else if (k == el) {
                count++;
            } else {
                count--;
            }
        }
        int cnt = 0;
        for (int j : arr) {
            if (j == el) {
                cnt++;
            }
        }

        if (cnt > arr.length / 2) {
            System.out.println(el);
        } else {
            System.out.println(-1);
        }
    }

    public static void main(String[] args) {
        int[] arr = {2, 2, 1, 3, 1, 1, 3, 1, 1, 3, 3, 3, 3, 3, 3, 1, 1, 1, 3, 3, 3, 3};
        majorityElement(arr);
    }
}
